clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -1238,52 +1422,33 @@ CreateOpenCLKernelDeclaration(ASTContext &Context, StringRef Name,
 //
 void Sema::ConstructOpenCLKernel(FunctionDecl *KernelCallerFunc,
                                  MangleContext &MC) {
-  CXXRecordDecl *LE = getKernelObjectType(KernelCallerFunc);
-  assert(LE && "invalid kernel caller");
-
-  // Build list of kernel arguments
-  llvm::SmallVector<ParamDesc, 16> ParamDescs;
-  if (!buildArgTys(getASTContext(), LE, ParamDescs))
-    return;
-
-  // Extract name from kernel caller parameters and mangle it.
-  const TemplateArgumentList *TemplateArgs =
-      KernelCallerFunc->getTemplateSpecializationArgs();
-  assert(TemplateArgs && "No template argument info");
-  QualType KernelNameType = TypeName::getFullyQualifiedType(
-      TemplateArgs->get(0).getAsType(), getASTContext(), true);
-
-  std::string Name;
-  // TODO SYCLIntegrationHeader also computes a unique stable name. It should
-  // probably lose this responsibility and only use the name provided here.
-  if (getLangOpts().SYCLUnnamedLambda)
-    Name = PredefinedExpr::ComputeName(
-        getASTContext(), PredefinedExpr::UniqueStableNameExpr, KernelNameType);
-  else
-    Name = constructKernelName(KernelNameType, MC);
-
-  // TODO Maybe don't emit integration header inside the Sema?
-  populateIntHeader(getSyclIntegrationHeader(), Name, KernelNameType, LE);
-
-  FunctionDecl *OpenCLKernel =
-      CreateOpenCLKernelDeclaration(getASTContext(), Name, ParamDescs);
-
-  ContextRAII FuncContext(*this, OpenCLKernel);
-
-  // Let's copy source location of a functor/lambda to emit nicer diagnostics
-  OpenCLKernel->setLocation(LE->getLocation());
-
-  // If the source function is implicitly inline, the kernel should be marked
-  // such as well. This allows the kernel to be ODR'd if there are multiple uses
-  // in different translation units.
-  OpenCLKernel->setImplicitlyInline(KernelCallerFunc->isInlined());
+  // The first argument to the KernelCallerFunc is the lambda object.
+  CXXRecordDecl *KernelLambda = getKernelObjectType(KernelCallerFunc);
+  assert(KernelLambda && "invalid kernel caller");
+
+  // Calculate both names, since Integration headers need both.
+  std::string CalculatedName =
+      constructKernelName(*this, KernelCallerFunc, MC, /*StableName*/ false);
+  std::string StableName =
+      constructKernelName(*this, KernelCallerFunc, MC, /*StableName*/ true);
```

Comment:
Does it make sense to return a pair of names from a single `constructKernelName` call?

Reply:
I think you're right, it does now.  Previously this was used separately in a couple of places that were removed as a part of the refactor.

Reply:
I guess, it should be fixed now.